### PR TITLE
[air/data] Concatenator preprocessor

### LINKED ISSUE #26500: [air] Add a Concat preprocessor
### Description

It would be useful to add a Concat(list_of_columns) preprocessor that returns the columns concatenated. This would simplify basic Torch and TF workflows: https://docs.ray.io/en/master/ray-air/getting-started.html#preprocessing

cc @richardliaw 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Adds a tensor concatenation preprocessor to make it easier to work with tabular data
## Related issue number

Closes https://github.com/ray-project/ray/issues/26500

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/examples/pytorch_tabular_starter.py [hunk 1]
```diff
@@ -33,27 +33,16 @@
 import numpy as np
 import pandas as pd
 
-from ray.data.preprocessors import BatchMapper, Chain
+from ray.data.preprocessors import BatchMapper, Chain, Tensorizer
```
Comment (ericl) [inline doc/source/ray-air/examples/pytorch_tabular_starter.py:36]: Is Tensorizer a common name? Would prefer something like Concatenate instead.

File: python/ray/data/preprocessors/tensorizer.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    Raises:
+        ValueError if any element in `columns` does not exist in the dataset.
+    """
+
+    _is_fittable = False
+
+    def __init__(
+        self,
+        columns: List[str],
+        output_column: str,
```
Comment (ericl) [inline python/ray/data/preprocessors/tensorizer.py:46]: Shall we set a default of "__value__" here to be consistent with single-column datasets?

File: doc/source/ray-air/examples/pytorch_tabular_starter.py [hunk 3]
```diff
@@ -33,27 +33,16 @@
...
-        target_data = dataframe["target"].to_numpy(dtype=np.float32)
-        result["target"] = TensorArray(target_data)
-    return pd.DataFrame(result)
-
-
 # Chain the preprocessors together.
-preprocessor = Chain(preprocessor, BatchMapper(concat_for_tensor))
+preprocessor = Chain(
+    preprocessor,
+    Tensorizer(columns=schema_order, output_column="input", dtype=np.float32),
```
Comment (ericl) [inline doc/source/ray-air/examples/pytorch_tabular_starter.py:44]:
```suggestion
    Concatenate(),
```

Let's avoid extra options in the basic examples. We can do this by having the default being to concat all columns.

File: doc/source/ray-air/examples/pytorch_tabular_starter.py [hunk 4]
```diff
@@ -33,27 +33,16 @@
 import numpy as np
 import pandas as pd
 
-from ray.data.preprocessors import BatchMapper, Chain
+from ray.data.preprocessors import BatchMapper, Chain, Tensorizer
 
 # Get the training data schema
 schema_order = [k for k in train_dataset.schema().names if k != "target"]
```
Comment (ericl) [inline doc/source/ray-air/examples/pytorch_tabular_starter.py:39]: Remove this from example.

Comment (richardliaw) [reply doc/source/ray-air/examples/pytorch_tabular_starter.py:39]: It will break though? See below comment.

**Comment by ericl** [comment]
How about we add an exclude option? exclude=["label"].

File: doc/source/ray-air/examples/pytorch_tabular_starter.py (hunk 1 above)
Comment (richardliaw) [reply doc/source/ray-air/examples/pytorch_tabular_starter.py]: fixed

File: python/ray/data/preprocessors/tensorizer.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/data/preprocessors/tensorizer.py]: This `VALUE_COL_NAME` isn't publicly documented anywhere right? Seems like we should first at least document it and move it out of `_internal` before using it?

Comment (ericl) [reply python/ray/data/preprocessors/tensorizer.py]: We should definitely document this (cc @clarkzinzow ). Btw, we can use it as a default arg without moving it out of internal, that's totally fine.

File: python/ray/data/preprocessors/concatenator.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class Concatenator(Preprocessor):
+    """Create tensor columns via concatenation.
+
+    A tensor column is a a column consisting of ndarrays as elements.
+
+    Example:
+        >>> import pandas as pd
+        >>> from ray.data.preprocessors import Concatenator
```
Comment (ericl) [inline python/ray/data/preprocessors/concatenator.py:16]:
```suggestion
        >>> from ray.data.preprocessors import Concatenate
```

Prefer the verb instead of the noun form for preprocessors.

File: python/ray/data/preprocessors/concatenator.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import pandas as pd
+
+from ray.data.extensions import TensorArray
+from ray.data.preprocessor import Preprocessor
+
+
+class Concatenator(Preprocessor):
+    """Create tensor columns via concatenation.
+
+    A tensor column is a a column consisting of ndarrays as elements.
```
Comment (ericl) [inline python/ray/data/preprocessors/concatenator.py:12]: Document what it does with the other columns (preserve)?

File: python/ray/data/preprocessors/tensorizer.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/preprocessors/tensorizer.py]: Actually, maybe better to use something like concat_out given we preserve other columns.

File: python/ray/data/tests/test_preprocessors.py [hunk 7]
```diff
@@ -1244,6 +1245,39 @@ def test_power_transformer():
...
+
+    with pytest.raises(ValueError, match="'b'"):
+        prep.transform(ds)
+
+    # check it works with string types
+    df = pd.DataFrame({"a": ["string", "string2", "string3"]})
+    ds = ray.data.from_pandas(df)
+    prep = Concatenator(output_column="huh")
+    new_ds = prep.transform(ds)
+    assert "huh" in df
```
Comment (ericl) [inline python/ray/data/tests/test_preprocessors.py:1278]: Test it preserves the original columns?

File: python/ray/data/preprocessors/concatenator.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    Raises:
+        ValueError if any element in `columns` does not exist in the dataset.
+    """
+
+    _is_fittable = False
+
+    def __init__(
+        self,
+        output_column: str,
+        exclude: Optional[List[str]] = None,
```
Comment (ericl) [inline python/ray/data/preprocessors/concatenator.py:45]: You should also have an "include" field (exclude shouldn't be the only way to do it).

File: python/ray/data/tests/test_preprocessors.py [hunk 9]
```diff
@@ -1244,6 +1245,39 @@ def test_power_transformer():
...
+    df = new_ds.to_pandas()
+    assert "c" in df
+    x = df["c"].iloc[0]
+    assert x.to_numpy().tolist() == [1, 1]
+
+    assert "c" in prep.__repr__()
+
+    df = pd.DataFrame({"a": [1, 2, 3, 4]})
+    ds = ray.data.from_pandas(df)
+    prep = Concatenator(output_column="c", exclude=["b"])
```
Comment (ericl) [inline python/ray/data/tests/test_preprocessors.py:1268]: We should also test exclude working, and that it preserves excluded columns.

File: python/ray/data/tests/test_preprocessors.py [hunk 10]
```diff
@@ -1244,6 +1245,39 @@ def test_power_transformer():
...
+    )
+    ds = ray.data.from_pandas(df)
+    prep = Concatenator(output_column="c")
+    new_ds = prep.transform(ds)
+    df = new_ds.to_pandas()
+    assert "c" in df
+    x = df["c"].iloc[0]
+    assert x.to_numpy().tolist() == [1, 1]
+
+    assert "c" in prep.__repr__()
```
Comment (ericl) [inline python/ray/data/tests/test_preprocessors.py:1264]: Instead of calling to_pandas, use take() and check the exact contents of the row outputs (that "c" is present and is [1, 2, 3, 4, 1, 2, 3, 4], and that "a" and "b" were deleted.

File: python/ray/data/preprocessors/tensorizer.py (hunk 2 above)
Comment (richardliaw) [reply python/ray/data/preprocessors/tensorizer.py]: OK, that's fine for the default. Will change. 

For the getting started example though, we shouldn't use this default value.

The trainer code later requires knowing what the keys are in order to index the iter_batches output properly, and having a __value__ magic keyword is very unintuitive.

Comment (richardliaw) [reply python/ray/data/preprocessors/tensorizer.py]: saw your comment below, sounds good.

File: python/ray/data/preprocessors/concatenator.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/data/preprocessors/concatenator.py:16]: Hmm, this would make it inconsistent with all of our other preprocessors.

Further, other ML frameworks (huggingface, scikit-learn, spark) have adopted the noun form for preprocessors too -- https://spark.apache.org/docs/latest/api/python/reference/pyspark.ml.html#feature

Comment (ericl) [reply python/ray/data/preprocessors/concatenator.py:16]: OK

File: python/ray/data/preprocessors/concatenator.py (hunk 8 above)
Comment (richardliaw) [reply python/ray/data/preprocessors/concatenator.py:49]: added

File: python/ray/data/tests/test_preprocessors.py (hunk 10 above)
Comment (richardliaw) [reply python/ray/data/tests/test_preprocessors.py:1263]: done, good tip.

File: doc/source/ray-air/examples/pytorch_tabular_starter.py [hunk 11]
```diff
@@ -114,7 +103,7 @@ def to_tensor_iterator(dataset, batch_size):
         session.report({"loss": loss}, checkpoint=to_air_checkpoint(model))
 
 
-num_features = len(schema_order)
+num_features = len(train_dataset.schema().names) - 1
 
 trainer = TorchTrainer(
     train_loop_per_worker=train_loop_per_worker,
```
Comment (ericl) [inline doc/source/ray-air/examples/pytorch_tabular_starter.py:109]: Could we remove the comments from the config below? They're mostly redundant.

File: doc/source/ray-air/examples/tf_tabular_starter.py [hunk 12]
```diff
@@ -149,7 +136,7 @@ def train_loop_per_worker(config):
         "batch_size": 128,
         # Number of epochs to train each task for.
         "num_epochs": 50,
-        # Number of columns of datset
+        # Number of columns of dataset
```
Comment (ericl) [inline doc/source/ray-air/examples/tf_tabular_starter.py:139]: Please remove these superfluous comments, here and elsewhere.

**Review by ericl — APPROVED** [review]
